From 222e8165907c8033075af7e6cc4c4413dade192b Mon Sep 17 00:00:00 2001
From: Vincent Brillault <git@lerya.net>
Date: Wed, 18 Nov 2015 20:01:30 +0100
Subject: [PATCH 01/17] Store details about successful auth methods

---
 auth.h    |  3 +++
 auth2.c   | 13 +++++++++++++
 session.c |  3 +++
 3 files changed, 19 insertions(+)

diff --git a/auth.h b/auth.h
index f9d191c..6c31f33 100644
--- a/auth.h
+++ b/auth.h
@@ -78,6 +78,9 @@ struct Authctxt {
 #endif
 	Buffer		*loginmsg;
 	void		*methoddata;
+
+	char		*last_details;
+	char		*auth_details;
 };
 /*
  * Every authentication method has to handle authentication requests for
diff --git a/auth2.c b/auth2.c
index 436cd60..b7ba160 100644
--- a/auth2.c
+++ b/auth2.c
@@ -306,6 +306,7 @@ userauth_finish(Authctxt *authctxt, int authenticated, const char *method,
     const char *submethod)
 {
 	char *methods;
+	char *prev_auth_details;
 	int partial = 0;
 
 	if (!authctxt->valid && authenticated)
@@ -336,6 +337,18 @@ userauth_finish(Authctxt *authctxt, int authenticated, const char *method,
 	if (authctxt->postponed)
 		return;
 
+	if (authenticated || partial) {
+		prev_auth_details = authctxt->auth_details;
+		xasprintf(&authctxt->auth_details, "%s%s%s%s%s",
+		    prev_auth_details ? prev_auth_details : "",
+		    prev_auth_details ? ", " : "", method,
+		    authctxt->last_details ? ": " : "",
+		    authctxt->last_details ? authctxt->last_details : "");
+		free(authctxt->last_details);
+		authctxt->last_details = NULL;
+		free(prev_auth_details);
+	}
+
 #ifdef USE_PAM
 	if (options.use_pam && authenticated) {
 		if (!PRIVSEP(do_pam_account())) {
diff --git a/session.c b/session.c
index 57c1f80..2988e29 100644
--- a/session.c
+++ b/session.c
@@ -2873,6 +2873,9 @@ do_cleanup(Authctxt *authctxt)
 	if (authctxt == NULL)
 		return;
 
+	free(authctxt->auth_details);
+	authctxt->auth_details = NULL;
+
 #ifdef USE_PAM
 	if (options.use_pam) {
 		sshpam_cleanup();
-- 
2.9.0

