From f8ea330ea73faa262b36fc5908cc6ce3f562458e Mon Sep 17 00:00:00 2001
From: Vincent Brillault <vincent.brillault@cern.ch>
Date: Tue, 23 Feb 2016 16:01:18 +0100
Subject: [PATCH 11/17] Privsep: expose detailed information for sshkey auth

---
 monitor.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/monitor.c b/monitor.c
index 794a970..62a1a8e 100644
--- a/monitor.c
+++ b/monitor.c
@@ -423,9 +423,13 @@ monitor_child_preauth(Authctxt *_authctxt, struct monitor *pmonitor)
 
 		if (authenticated) {
 			prev_auth_details = authctxt->auth_details;
-			xasprintf(&authctxt->auth_details, "%s%s%s",
+			xasprintf(&authctxt->auth_details, "%s%s%s%s%s",
 			    prev_auth_details ? prev_auth_details : "",
-			    prev_auth_details ? ", " : "", auth_method);
+			    prev_auth_details ? ", " : "", auth_method,
+			    authctxt->last_details ? ": " : "",
+			    authctxt->last_details ? authctxt->last_details : "");
+			free(authctxt->last_details);
+			authctxt->last_details = NULL;
 			free(prev_auth_details);
 		}
 
@@ -1482,6 +1486,9 @@ mm_answer_keyverify(int sock, Buffer *m)
 	debug3("%s: key %p signature %s",
 	    __func__, key, (verified == 1) ? "verified" : "unverified");
 
+	if (verified == 1)
+		authctxt->last_details = key_format_oneline(key);
+
 	key_free(key);
 	free(blob);
 	free(signature);
-- 
2.9.0

