From 67a1d106a7274a8eb692ff8dfdcdec15a4114950 Mon Sep 17 00:00:00 2001
From: Vincent Brillault <vincent.brillault@cern.ch>
Date: Tue, 23 Feb 2016 17:39:37 +0100
Subject: [PATCH 13/17] Expose displayname for gssapi logins

---
 auth2-gss.c | 6 ++++++
 monitor.c   | 3 +++
 2 files changed, 9 insertions(+)

diff --git a/auth2-gss.c b/auth2-gss.c
index ad65059..6089d20 100644
--- a/auth2-gss.c
+++ b/auth2-gss.c
@@ -272,6 +272,9 @@ input_gssapi_exchange_complete(int type, u_int32_t plen, void *ctxt)
 	authenticated = PRIVSEP(ssh_gssapi_userok(authctxt->user,
 	    authctxt->pw));
 
+	if (authenticated)
+		authctxt->last_details = ssh_gssapi_get_displayname();
+
 	authctxt->postponed = 0;
 	dispatch_set(SSH2_MSG_USERAUTH_GSSAPI_TOKEN, NULL);
 	dispatch_set(SSH2_MSG_USERAUTH_GSSAPI_ERRTOK, NULL);
@@ -317,6 +320,9 @@ input_gssapi_mic(int type, u_int32_t plen, void *ctxt)
 	else
 		logit("GSSAPI MIC check failed");
 
+	if (authenticated)
+		authctxt->last_details = ssh_gssapi_get_displayname();
+
 	buffer_free(&b);
 	if (micuser != authctxt->user)
 		free(micuser);
diff --git a/monitor.c b/monitor.c
index 62a1a8e..ca61729 100644
--- a/monitor.c
+++ b/monitor.c
@@ -2338,6 +2338,9 @@ mm_answer_gss_userok(int sock, Buffer *m)
 
 	auth_method = "gssapi-with-mic";
 
+	if (authenticated)
+                authctxt->last_details = ssh_gssapi_get_displayname();
+
 	/* Monitor loop will terminate if authenticated */
 	return (authenticated);
 }
-- 
2.9.0

