From 4fefeb695acda04f52b5ec5672a9c3e70dcc9ce9 Mon Sep 17 00:00:00 2001
From: Vincent Brillault <vincent.brillault@cern.ch>
Date: Fri, 1 Jul 2016 09:04:45 +0200
Subject: [PATCH 17/17] ExposeAuthenticationMethods: change default to 'never'

---
 servconf.c    | 2 +-
 sshd_config.5 | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/servconf.c b/servconf.c
index 4512024..d339276 100644
--- a/servconf.c
+++ b/servconf.c
@@ -338,7 +338,7 @@ fill_default_server_options(ServerOptions *options)
 #endif
 
 	if (options->expose_auth_methods == -1)
-		options->expose_auth_methods = EXPOSE_AUTHMETH_PAMONLY;
+		options->expose_auth_methods = EXPOSE_AUTHMETH_NEVER;
 }
 
 /* Keyword tokens. */
diff --git a/sshd_config.5 b/sshd_config.5
index 0a5eb24..429489a 100644
--- a/sshd_config.5
+++ b/sshd_config.5
@@ -475,7 +475,7 @@ Valid options are:
 .Dq pam-and-env
 (Expose them to PAM and keep them in the shell environment).
 The default is
-.Dq pam-only .
+.Dq never .
 .It Cm ForceCommand
 Forces the execution of the command specified by
 .Cm ForceCommand ,
-- 
2.9.0

