From 312cad143a1c04b9fd556e0937ca086f2fa68a4c Mon Sep 17 00:00:00 2001
From: Vincent Brillault <vincent.brillault@cern.ch>
Date: Tue, 1 Dec 2015 15:04:18 +0100
Subject: [PATCH 2/5] Export auth_details to child env as SSH_USER_AUTH

---
 session.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/session.c b/session.c
index cf7fbd5..2c90943 100644
--- a/session.c
+++ b/session.c
@@ -1365,6 +1365,10 @@ do_setup_env(Session *s, const char *shell)
 	}
 #endif /* USE_PAM */
 
+	if (s->authctxt->auth_details)
+		child_set_env(&env, &envsize, "SSH_USER_AUTH",
+		     s->authctxt->auth_details);
+
 	if (auth_sock_name != NULL)
 		child_set_env(&env, &envsize, SSH_AUTHSOCKET_ENV_NAME,
 		    auth_sock_name);
-- 
1.8.3.1

